tests/mgbench/benchmark.py

```diff
@@ -144,45 +104,427 @@ def filter_benchmarks(generators, patterns):
+                        variant,
+                        group,
+                        query_name,
+                        is_default_variant,
+                        patterns,
+                    ):
+                        current[group].append((query_name, query_func))
+            if len(current) > 0:
+                filtered.append((generator(variant), dict(current)))
+    return filtered
+
+
+def warmup(client):
+    print("Executing warm-up queries")
+    client.execute(
+        queries=[
+            ("CREATE ();", {}),
+            ("CREATE ()-[:TempEdge]->();", {}),
+            ("MATCH (n) RETURN n LIMIT 1;", {}),
+        ],
+        num_workers=1,
+    )
+
+
+def tail_latency(vendor, client, func):
+    vendor.start_benchmark()
+    if args.warmup_run:
+        warmup(client)
+    latency = []
+    iteration = 100
+    query_list = get_queries(func, 100)
+    for i in range(0, iteration):
+        ret = client.execute(queries=[query_list[i]], num_workers=1)
+        latency.append(ret[0]["duration"])
+    latency.sort()
+    query_stats = {
+        "iterations": iteration,
+        "min": latency[0],
+        "max": latency[99],
+        "mean": statistics.mean(latency),
+        "p99": latency[98],
+        "p95": latency[94],
+        "p90": latency[89],
+        "p75": latency[74],
+        "p50": latency[49],
+    }
+    print("Query statistics for tail latency: ")
+    print(query_stats)
+    vendor.stop()
+    return query_stats
+
+
+def mixed_workload(vendor, client, dataset, group, queries):
+    num_args = len(args.mixed_workload)
+    if num_args > 6:
+        raise Exception(
+            "You have passed to many arguments for mixed workload",
+            "read description of --mixed-worload args",
+        )
+
+    num_of_queries = args.mixed_workload[0]
+    percentage_distribution = args.mixed_workload[1:]
+    if sum(percentage_distribution) != 100:
+        raise Exception(
+            "Please make sure that passed arguments % sum to 100% percent!, passed: ",
+            percentage_distribution,
+        )
+    s = [str(i) for i in args.mixed_workload]
+
+    config_distribution = "_".join(s)
+
+    print("Generating mixed workload")
+    write = list()
+    read = list()
+    update = list()
+    analytical = list()
+
+    for (
+        query,
+        funcname,
+    ) in queries[group]:
+        if "write" in funcname:
+            write.append(funcname)
+        elif "read" in funcname:
+            read.append(funcname)
+        elif "update" in funcname:
+            update.append(funcname)
+        elif "analytical" in funcname:
+            analytical.append(funcname)
+
+    if (
+        len(write) == 0
+        and percentage_distribution[0] != 0
+        or len(read) == 0
+        and percentage_distribution[1] != 0
+        or len(update) == 0
+        and percentage_distribution[2] != 0
+        or len(analytical) == 0
+        and percentage_distribution[3] != 0
```

Comment:
Ideas:
* I would write a function for a single check (`len(write) == 0 and distribution != 0`) and then use https://docs.python.org/3/library/functions.html#any
* This whole block of code (list init + for + this if) seems very repetitive and error-prone + since there are `ops`, `distributions`, maybe there is a nice/short abstraction for that? But that might be too much for this simple piece of logic here...

Reply:
Regarding the cleaning of repetitive code, I cleaned it up by using dictionaries.